Create get-started-windows-instructions.md (#16465)
dotnet/docs: docs/spark/tutorials/get-started-windows-instructions.md

---

> title: Get Started with .NET for Apache Spark on Windows

Reviewer: ```suggestion
title: Get started with .NET for Apache Spark on Windows
```

description: Discover how to run a .NET for Apache Spark app using .NET Core on Windows.
ms.date: 12/30/2019
ms.topic: tutorial
ms.custom: mvc
# Customer intent: As a developer, I want to write a simple custom application using .NET for Apache Spark on Windows.
---

> # Tutorial: Get Started with .NET for Apache Spark on Windows

Reviewer: ```suggestion
# Tutorial: Get started with .NET for Apache Spark on Windows
```

This tutorial will show you how to run a .NET for Apache Spark app using .NET Core on Windows.

> [!div class="checklist"]
>
> * Pre-requisites to run your .NET for Apache Spark application
> * Write a simple .NET for Apache Spark application
> * Build and run your .NET for Apache Spark application

## Pre-requisites

> - Download and install the following: **[.NET Core 2.1 SDK](https://dotnet.microsoft.com/download/dotnet-core/2.1)** | **[Visual Studio 2019](https://www.visualstudio.com/downloads/)** | **[Java 1.8](https://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html)** | **[Apache Spark 2.4.1](https://archive.apache.org/dist/spark/spark-2.4.1/spark-2.4.1-bin-hadoop2.7.tgz)**

Reviewer 2: either say following products or just list one by one

also links to VS downloads from docs should use campaign links
https://visualstudio.microsoft.com/downloads/?utm_medium=microsoft&utm_source=docs.microsoft.com&utm_campaign=inline+link&utm_content=download+vs2019

Author: Thanks for your comments! I have changed to `following products`. 
Can you please specify what you mean by `campaign links`? Thanks.

- Download and install **[Microsoft.Spark.Worker](https://github.com/dotnet/spark/releases)** release:

>   - Select a **[Microsoft.Spark.Worker](https://github.com/dotnet/spark/releases)** release from .NET for Apache Spark GitHub Releases page and download into your local machine (e.g., `c:\bin\Microsoft.Spark.Worker\`).

Reviewer 2: https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/eg

```suggestion
  - Select a **[Microsoft.Spark.Worker](https://github.com/dotnet/spark/releases)** release from .NET for Apache Spark GitHub Releases page and download into your local machine (for example, `c:\bin\Microsoft.Spark.Worker\`).
```

>   - **IMPORTANT** Create a [new environment variable](https://www.java.com/en/download/help/path.xml) `DOTNET_WORKER_DIR` and set it to the directory where you downloaded and extracted the Microsoft.Spark.Worker (e.g., `c:\bin\Microsoft.Spark.Worker`).

Reviewer: You can use the [!IMPORTANT] syntax.

Directories are italics, based on current style guide

Reviewer 2: You would have to check the exact positioning of this text but this is what @Youssef1313 meant
```suggestion

  > [!IMPORTANT]
 > Create a [new environment variable](https://www.java.com/en/download/help/path.xml) `DOTNET_WORKER_DIR` and set it to the directory where you downloaded and extracted the Microsoft.Spark.Worker (for example, *c:\bin\Microsoft.Spark.Worker*).
```

> For detailed instructions, you can see [Building .NET for Apache Spark from Source on Windows](../building/windows-instructions.md).

Reviewer 2: this doesn't exist

Author: Changed. Thanks. This link should be a doc from how-to-guides session and it's currently an open pr.

## Authoring a .NET for Apache Spark App

> - Open Visual Studio -> Create New Project -> Console App (.NET Core) -> Name: `HelloSpark`

Reviewer: ```suggestion
1. Open Visual Studio > **Create New Project** > **Console App (.NET Core)** > Name: `HelloSpark`.
```

Reviewer 2: These instructions seem to take a shortcut. Advanced users will probably know what you mean but not someone that is getting started.

Here's an example of a Get started with VS:
https://docs.microsoft.com/en-us/dotnet/core/tutorials/with-visual-studio

Author: Thanks for your suggestion! I've added a link.

> - Install `Microsoft.Spark` Nuget package into the solution from the [spark nuget.org feed](https://www.nuget.org/profiles/spark) - see [Ways to install Nuget Package](https://docs.microsoft.com/en-us/nuget/consume-packages/ways-to-install-a-package)

Reviewer: ```suggestion
1. Install `Microsoft.Spark` Nuget package into the solution from the [spark nuget.org feed](https://www.nuget.org/profiles/spark) - see [Ways to install Nuget Package](https://docs.microsoft.com/nuget/consume-packages/ways-to-install-a-package)
```

> - Write the following code into `Program.cs`:

Reviewer: ```suggestion
1. Write the following code into *Program.cs*:
```

    ```csharp
    using Microsoft.Spark.Sql;

    namespace HelloSpark
    {
        class Program
        {
            static void Main(string[] args)
            {

>                 var spark = SparkSession.Builder().GetOrCreate();

Reviewer: I 'd use actual data type, not var.

Author: Changed to the actual data type. Thanks!

                var df = spark.Read().Json("people.json");
                df.Show();
            }
        }
    }
    ```

> - Build the solution

Reviewer: ```suggestion
1. Build the solution
```

## Running your .NET for Apache Spark App

> - Open your terminal and navigate into your app folder:

Reviewer: ```suggestion
1. Open your terminal and navigate into your app folder:
```

>     ```

Reviewer: language identifier needed here. Not sure if console or bash is the suitable

Author: I would say `batch`, but not sure if this is accurate. @mamccrea any suggestion?

Reviewer 3: You're using `bash` below.

    cd <your-app-output-directory>
    ```
    

> - Create `people.json` with the following content:

Reviewer: ```suggestion
1. Create *people.json* with the following content:
```

    ```json
    {"name":"Michael"}
    {"name":"Andy", "age":30}
    {"name":"Justin", "age":19}
    ```
    

> - Run your app

Reviewer: ```suggestion
1. Run your app
```

>     ```

Reviewer: language identifier ?

Author: Added. Thanks!

    spark-submit `
    --class org.apache.spark.deploy.dotnet.DotnetRunner `
    --master local `
    microsoft-spark-2.4.x-<version>.jar `
    dotnet HelloSpark.dll
    ```
    

>     **Note**: This command assumes you have downloaded Apache Spark and added it to your PATH environment variable to be able to use `spark-submit`, otherwise, you would have to use the full path (e.g., `c:\bin\apache-spark\bin\spark-submit`). For detailed instructions, you can see [Building .NET for Apache Spark from Source on Windows](../building/windows-instructions.md).

Reviewer: Use alert-style note

Reviewer 2: ```suggestion
    > [!NOTE]
    > This command assumes you have downloaded Apache Spark and added it to your PATH environment variable to be able to use `spark-submit`. Otherwise, you'd have to use the full path (for example, *c:\bin\apache-spark\bin\spark-submit*). For detailed instructions, see [Building .NET for Apache Spark from Source on Windows](../building/windows-instructions.md).
```

> - The output of the application should look similar to the output below:

Reviewer: ```suggestion
1. The output of the application should look similar to the output below:
```

Reviewer 2: output below -> following output

https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/b/below

Author: Changed. Thanks!

    ```text
    +----+-------+
    | age|   name|
    +----+-------+
    |null|Michael|
    |  30|   Andy|
    |  19| Justin|
    +----+-------+
    ```
    
## Next steps

> In this tutorial, you learn how to run your .NET for Apache Spark on Windows. Continue to learn more about how to run your .NET for Apache Spark on Ubuntu and MacOSX.

Reviewer 2: ```suggestion
In this tutorial, you learn how to run your .NET for Apache Spark on Windows. Continue to learn more about how to run your .NET for Apache Spark on Ubuntu and macOS.

```

> [!div class="nextstepaction"]

> > [Tutorial: Get Started with .NET for Apache Spark on Ubuntu](get-started-ubuntu-instructions.md)

Reviewer 2: if the three new articles are in separate PRs, maybe we should create a release branch for this content so this only gets merged to master when all three are ready and free of warnings.

Author: Could you suggest me how to create a release branch for these three PRs?

---

# Tutorial: Get Started with .NET for Apache Spark on Windows

This tutorial will show you how to run a .NET for Apache Spark app using .NET Core on Windows.

> [!div class="checklist"]
>
> * Pre-requisites to run your .NET for Apache Spark application
> * Write a simple .NET for Apache Spark application
> * Build and run your .NET for Apache Spark application

## Pre-requisites

- Download and install the following: **[.NET Core 2.1 SDK](https://dotnet.microsoft.com/download/dotnet-core/2.1)** | **[Visual Studio 2019](https://www.visualstudio.com/downloads/)** | **[Java 1.8](https://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html)** | **[Apache Spark 2.4.1](https://archive.apache.org/dist/spark/spark-2.4.1/spark-2.4.1-bin-hadoop2.7.tgz)**
- Download and install **[Microsoft.Spark.Worker](https://github.com/dotnet/spark/releases)** release:
  - Select a **[Microsoft.Spark.Worker](https://github.com/dotnet/spark/releases)** release from .NET for Apache Spark GitHub Releases page and download into your local machine (e.g., `c:\bin\Microsoft.Spark.Worker\`).

>   > [!IMPORTANT]

Reviewer 4: ```suggestion
 > [!IMPORTANT]
 >
```

Add space to fix lint

description: Discover how to run a .NET for Apache Spark app using .NET Core on Windows.
ms.date: 01/22/2019
ms.topic: tutorial
ms.custom: mvc
# Customer intent: As a developer, I want to write a simple custom application using .NET for Apache Spark on Windows.
---

# Tutorial: Get started with .NET for Apache Spark on Windows

This tutorial will show you how to run a .NET for Apache Spark app using .NET Core on Windows.

> [!div class="checklist"]
>
> * Pre-requisites to run your .NET for Apache Spark application
> * Write a simple .NET for Apache Spark application
> * Build and run your .NET for Apache Spark application

## Pre-requisites

- Download and install the following products: **[.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet-core/3.1)** | **[Visual Studio 2019](https://www.visualstudio.com/downloads/)** | **[Java 1.8](https://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html)** | **[Apache Spark 2.4.1](https://archive.apache.org/dist/spark/spark-2.4.1/spark-2.4.1-bin-hadoop2.7.tgz)**
- Download and install **[Microsoft.Spark.Worker](https://github.com/dotnet/spark/releases)** release:

>   - Select a **[Microsoft.Spark.Worker](https://github.com/dotnet/spark/releases)** release from .NET for Apache Spark GitHub Releases page and download into your local machine (for example, `c:\bin\Microsoft.Spark.Worker\`).

Author: Good point! Changed it to upper-case `C`.

 > [!IMPORTANT]
 >
 > Create a [new environment variable](https://www.java.com/en/download/help/path.xml) `DOTNET_WORKER_DIR` and set it to the directory where you downloaded and extracted the Microsoft.Spark.Worker (for example, *c:\bin\Microsoft.Spark.Worker*).

For detailed instructions, you can see [Building .NET for Apache Spark from Source on Windows](../how-to-guides/windows-instructions.md).

## Authoring a .NET for Apache Spark App

1. Open Visual Studio > **Create New Project** > **Console App (.NET Core)** > Name: `HelloSpark`.

>    For detailed instructions, you can see [How to create the app in Visual Studio](https://docs.microsoft.com/en-us/dotnet/core/tutorials/with-visual-studio?tabs=csharp).

Reviewer: The existence of en-us causes build warning. I even think it's better to use relative link here.
I'm suggesting also to remove the tabs=csharp to keep the last choice of the user.

```suggestion
   For detailed instructions, you can see [How to create the app in Visual Studio](../../core/tutorials/with-visual-studio.md).
```

Author: Sounds good! I've changed accordingly. Thanks!

1. Write the following code into *Program.cs*:

    ```csharp
    using Microsoft.Spark.Sql;

    namespace HelloSpark
    {
        class Program
        {
            static void Main(string[] args)
            {
                SparkSession spark = SparkSession.Builder().GetOrCreate();
                DataFrame df = spark.Read().Json("people.json");
                df.Show();
            }
        }
    }
    ```

1. Build the solution

## Running your .NET for Apache Spark App

1. Open your terminal and navigate into your app folder:

    ```
    cd <your-app-output-directory>
    ```
    
1. Create *people.json* with the following content:

    ```json
    {"name":"Michael"}
    {"name":"Andy", "age":30}
    {"name":"Justin", "age":19}
    ```
    
1. Run your app

    ```batch
    spark-submit `
    --class org.apache.spark.deploy.dotnet.DotnetRunner `
    --master local `
    microsoft-spark-2.4.x-<version>.jar `
    dotnet HelloSpark.dll

>     ```

Reviewer: I think you meant bash ? Unless there is a `batch` identifier that I'm not aware of.
I also don't understand the back tick at the end of each line. I suspect it should be a backslash to escape the new line ?

Author: Yes, changed it to bash and added the backslash to escape newline. Thanks!

    
    > [!NOTE]

>     > This command assumes you have downloaded Apache Spark and added it to your PATH environment variable to be able to use `spark-submit`. Otherwise, you'd have to use the full path (for example, *c:\bin\apache-spark\bin\spark-submit*). For detailed instructions, see [Building .NET for Apache Spark from Source on Windows](../building/windows-instructions.md).

Reviewer 3: Do we want this warning to be below the command?  I wonder if a reader might not see it.

Author: With the [!NOTE] rending, I think it's kind obvious. Please refer to warning session [here](https://docs.microsoft.com/en-us/dotnet/spark/how-to-guides/deploy-worker-udf-binaries#parameter-options).
